apps/mobile/src/hooks/useDelegateCleanup/utils.ts

```diff
@@ -0,0 +1,311 @@
+import Logger from '@/src/utils/logger'
+import { type Address } from '@/src/types/address'
+import { Wallet } from 'ethers'
+import { getDelegateTypedData } from '@safe-global/utils/services/delegates'
+import { type Chain } from '@safe-global/store/gateway/AUTO_GENERATED/chains'
+import { type DelegatesDeleteDelegateV2ApiArg } from '@safe-global/store/gateway/AUTO_GENERATED/delegates'
+import { keyStorageService } from '@/src/services/key-storage'
+import { getDelegateKeyId } from '@/src/utils/delegate'
+
+// Types for cleanup results
+interface CleanupResult {
+  success: boolean
+  error?: {
+    message: string
+    code?: string
+  }
+}
+
+interface NotificationCleanupResult {
+  success: boolean
+  error?: string
+  failedDelegates?: Address[]
+}
+
+interface DelegateRemovalResult {
+  success: boolean
+  error?: string
+  failedDelegates?: Address[]
+}
+
+interface KeychainCleanupResult {
+  success: boolean
+  error?: string
+  failedDelegates?: Address[]
+}
+
+/**
+ * Cleans up notifications for all delegates of a given owner
+ * This is a critical step that must succeed before proceeding with delegate removal
+ */
+export const cleanupDelegateNotifications = async (
+  ownerAddress: Address,
+  delegateAddresses: Address[],
+  cleanupNotificationsForDelegate: (ownerAddress: Address, delegateAddress: Address) => Promise<CleanupResult>,
+): Promise<NotificationCleanupResult> => {
+  if (!delegateAddresses || delegateAddresses.length === 0) {
+    return { success: true }
+  }
+
+  try {
+    const notificationCleanupResults = await Promise.allSettled(
+      delegateAddresses.map(async (delegateAddress) => {
+        try {
+          const result = await cleanupNotificationsForDelegate(ownerAddress, delegateAddress)
+          if (!result.success && result.error) {
+            throw new Error(`Notification cleanup failed for ${delegateAddress}: ${result.error.message}`)
+          }
+          return result
+        } catch (error) {
+          Logger.error(`Failed to cleanup notifications for delegate ${delegateAddress}`, error)
+          throw error
+        }
+      }),
+    )
+
+    // Check if any notification cleanup failed with blocking errors
+    const failedCleanups = notificationCleanupResults
+      .filter((result) => result.status === 'rejected')
+      .map((result) => result.reason)
+
+    if (failedCleanups.length > 0) {
+      const failedDelegates = delegateAddresses.filter(
+        (_, index) => notificationCleanupResults[index].status === 'rejected',
+      )
+
+      const errorMsg = `Cannot delete private key: ${failedCleanups.join(', ')}. Please check your internet connection and try again.`
```

Comment:
Seems a bit strange to me that we don't allow users to delete their signer unless notifications/delegates are also deleted. I understand the reason why we do that but IMO notifications should not be a reason to block something as critical as removing your private key from an app.